core/utils/historical-data/src/linear.rs

```diff
@@ -0,0 +1,447 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Linear historical data.
+
+#[cfg(not(feature = "std"))]
+use rstd::{vec::Vec, vec};
+use rstd::marker::PhantomData;
+use rstd::borrow::Cow;
+use crate::HistoricalValue;
+
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(not(feature = "std"))]
+pub(crate) type MemoryOnly<V, I> = Vec<HistoricalValue<V, I>>;
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(feature = "std")]
+pub(crate) type MemoryOnly<V, I> = smallvec::SmallVec<[HistoricalValue<V, I>; ALLOCATED_HISTORY]>;
+
+/// Size of preallocated history per element.
+/// Currently at two for committed and prospective only.
+/// It means that using transaction in a module got a direct allocation cost.
+#[cfg(feature = "std")]
+const ALLOCATED_HISTORY: usize = 2;
+
+#[derive(Debug, Clone)]
+#[cfg_attr(any(test, feature = "test"), derive(PartialEq))]
+/// Arraylike buffer with in place byte data.
+/// Can be written as is in underlying
+/// storage.
+/// Could be extended to direct access memory too.
+pub struct Serialized<'a, F>(SerializedBuff<'a>, PhantomData<F>);
+
+#[derive(Debug)]
+#[cfg_attr(any(test, feature = "test"), derive(PartialEq))]
+enum SerializedBuff<'a> {
```

Comment:
doc ?

also can't you just have the variant copy on write and just copy it on writes ?

Reply:
like `Cow<Vec<u8>>` ?

Reply:
This is indeed looking a bit odd, I am not totally sure if it is possible. In my first implementation it was a simple 'cow<[u8]>', but I probably reach a case where my handle was an '&mut Vec<u8>' and reach the conclusion that I would not want to go through cloning and replacing the buffer afterward. In practice this is only use for InMemory case (direct change of the value in the hashmap), but not with rocksdb.

Reply:
isn't the funciton [to_mut](file:///home/thiolliere/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/share/doc/rust/html/std/borrow/enum.Cow.html#method.to_mut) doing already the job ?

Reply:
Yes, probably need to store some Cow in the InMemory Hashmap then (I do not know a good way to handle hashmap 'get_mut' otherwhise).